x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/support/StringMatcher.java

```diff
@@ -0,0 +1,160 @@
+ */
+
+package org.elasticsearch.xpack.core.security.support;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.util.automaton.TooComplexToDeterminizeException;
+import org.elasticsearch.ElasticsearchSecurityException;
+import org.elasticsearch.common.Strings;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+/**
+ * This class acts as a facade / encapsulation around the expression and testing of string-based patterns within Elasticsearch security.
+ * Security supports "wildcards" in a number of places (e.g. index names within roles). These cases also support
+ * {@link org.apache.lucene.util.automaton.RegExp Lucene-syntax regular expressions} and are implemented via Lucene
+ * {@link org.apache.lucene.util.automaton.Automaton} objects.
+ * However, it can be more efficient to have special handling and avoid {@code Automata} for particular cases such as exact string matches.
+ * This class handles that logic, an provides a clean interface for
+ * <em>test whether a provided string matches one of an existing set of patterns</em> that hides the possible implementation options.
+ */
+public class StringMatcher implements Predicate<String> {
+
+    private static final StringMatcher MATCH_NOTHING = new StringMatcher("(empty))", s -> false);
+
+    private final String description;
+    private final Predicate<String> predicate;
+    private static final Logger LOGGER = LogManager.getLogger(StringMatcher.class);
+
+    private StringMatcher(String description, Predicate<String> predicate) {
+        this.description = description;
+        this.predicate = predicate;
+    }
+
+    public static StringMatcher of(Iterable<String> patterns) {
+        return StringMatcher.builder().includeAll(patterns).build();
+    }
+
+    public static StringMatcher of(String... patterns) {
+        return StringMatcher.builder().includeAll(patterns).build();
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    @Override
+    public String toString() {
+        return description;
+    }
+
+    @Override
+    public boolean test(String s) {
+        return predicate.test(s);
+    }
+
+    public StringMatcher or(Predicate<? super String> other) {
+        Objects.requireNonNull(other);
+        return new StringMatcher(description + "|" + other, this.predicate.or(other));
+    }
+
+    public StringMatcher and(Predicate<? super String> other) {
+        return this.and(String.valueOf(other), other);
+    }
+
+    public StringMatcher and(String otherDescription, Predicate<? super String> otherPredicate) {
+        Objects.requireNonNull(otherPredicate);
+        return new StringMatcher(this.description + "&" + otherDescription, this.predicate.and(otherPredicate));
+    }
```

Comment:
Should these methods have argument of type `StringMatcher` instead of `Predicate`? A raw predicate does not make much sense to have `description`?
Nits:
* It seems the two `and` methods can be merged into one if it follows the pattern of the `or` method.
* If any of the description is itself a combined one, e.g. `a|b|c`, when concatenating it again with another description, would it be helpful to put parenthesis around it, e.g. `(a|b|c)&d`?

Reply:
@ywangd I believe this is Predicate and and or default methods signature we have to use. We can build the description from the Predicate.

Reply:
These methods are not annotated with `@Override`. So I thought the intention was to have different methods instead of the default ones. 
With that being said, Java does not have multiple dispatch. A method of `public StringMatcher and(StringMatcher)` does not guarantee to work as intended all the time. So its usefulness is arguable. Other options could be a new interface extending `Predicate` or assert the argument's type. But they are not perfect either.

Reply:
The first two should be marked `@Override` (I think IntelliJ does weird things, with overriding default implementations...).
The 3rd (`and(String,Predicate)`) isn't an override, but is intentionally a `Predicate` so that you can apply a description to a generic predicate. It gets used in `IndicesPermission` for that exact purpose.